qe_metrics/utils/issue_utils.py

```diff
@@ -0,0 +1,110 @@
+from datetime import datetime, date
+from pony import orm
+from typing import List
+from jira import Issue
+from qe_metrics.libs.database_mapping import ProductsEntity, JiraIssuesEntity
+from simple_logger.logger import get_logger
+
+LOGGER = get_logger(name=__name__)
+
+
+def format_issue_date(date_str: str) -> date:
+    """
+    Format the date in the format YYYY-MM-DD.
+
+    Args:
+        date_str (str): Date string from Jira.Issue
+
+    Returns:
+        date: Date object for database insertion
+    """
+    return datetime.strptime(date_str, "%Y-%m-%dT%H:%M:%S.%f%z").date()
+
+
+def mark_obsolete_issues(
+    current_issues: List[Issue],
+    db_issues: List["JiraIssuesEntity"],
+    product: "ProductsEntity",  # type: ignore  # noqa: F821
+) -> None:
+    """
+    Mark JiraIssuesEntity items as "obsolete" in the database that are not in the current list of issues.
+
+    Args:
+        current_issues (List[Issue]): A list of current Jira issues for a product.
+        db_issues ("JiraIssuesEntity"): A list of JiraIssuesEntity objects already in the database.
+        product ("ProductsEntity"): The product object to which the issues belong.
+    """
+    current_issue_keys = {issue.key for issue in current_issues}
+    db_issue_keys = {db_issue.issue_key for db_issue in db_issues if db_issue.product.name == product.name}
+    closed_issue_keys = db_issue_keys - current_issue_keys
+
+    for db_issue in db_issues:
+        if db_issue.issue_key in closed_issue_keys and db_issue.status != "obsolete":
+            db_issue.logger.info(f'Marking issue "{db_issue.issue_key}" for product {product.name} as obsolete')
+            db_issue.status = "obsolete"
+
+
+def update_existing_issue(existing_issue: JiraIssuesEntity, new_issue_data: Issue, severity: str) -> None:
+    """
+    Check if any of the fields of an existing issue have changed and update them if they have.
+
+    Args:
+        existing_issue (JiraIssuesEntity): Existing issue from the database
+        issue (Issue): New issue from Jira
+        severity (str): Severity assigned to the issue/query
+    """
+    fields = [
+        ("title", new_issue_data.fields.summary.strip()),
+        ("severity", severity),
+        ("status", new_issue_data.fields.status.name),
+        ("customer_escaped", new_issue_data.is_customer_escaped),
+        ("last_updated", format_issue_date(new_issue_data.fields.updated)),
+    ]
+    for field, new_value in fields:
+        if (current_value := getattr(existing_issue, field)) != new_value:
+            LOGGER.info(
+                f'Updating issue "{new_issue_data.key}" in database: "{field}" changed from "{current_value}" to "{new_value}"'
+            )
+            setattr(existing_issue, field, new_value)
+
+
+def create_update_issues(
+    issues: List[Issue],
+    product: "ProductsEntity",
+    severity: str,
+    jira_server: str,  # type: ignore  # noqa: F821
```

Comment:
`# type: ignore  # noqa: F821` why this is needed?

Reply:
Must have been leftover from something else. Removed.